src/index/CompressedRelation.cpp

```diff
@@ -312,80 +341,140 @@ bool CompressedRelationReader::FilterDuplicatesAndGraphs::canBlockBeSkipped(
-        scanSpec, config, *it, std::ref(details), locatedTriplesPerBlock);
-    cancellationHandle->throwIfCancelled();
-    return result;
-  };
+    auto getIncompleteBlock(std::vector<CompressedBlockMetadata>::iterator it) {
+      auto result = reader->readPossiblyIncompleteBlock(
+          scanSpec, config, *it, std::ref(details()), locatedTriplesPerBlock);
+      cancellationHandle->throwIfCancelled();
+      return result;
+    };
 
-  // Get and yield the first block.
-  if (beginBlockMetadata < endBlockMetadata) {
-    auto block = getIncompleteBlock(beginBlockMetadata);
-    pruneBlock(block, limitOffset);
-    details.numElementsYielded_ += block.numRows();
-    if (!block.empty()) {
-      co_yield block;
-    }
-  }
+    std::optional<IdTable> get() override {
+      if (!firstBlockYielded) {
+        firstBlockYielded = true;
+        // Get and yield the first block.
+        if (beginBlockMetadata < endBlockMetadata) {
+          auto block = getIncompleteBlock(beginBlockMetadata);
+          pruneBlock(block, limitOffset);
+          details().numElementsYielded_ += block.numRows();
+          if (!block.empty()) {
+            return block;
+          }
+        }
+      }
 
-  // Get and yield the remaining blocks.
-  if (beginBlockMetadata + 1 < endBlockMetadata) {
-    // We copy the cancellationHandle because it is still captured by
-    // reference inside the `getIncompleteBlock` lambda.
-    auto blockGenerator = asyncParallelBlockGenerator(
-        beginBlockMetadata + 1, endBlockMetadata - 1, config,
-        cancellationHandle, limitOffset);
-    blockGenerator.setDetailsPointer(&details);
-    for (auto& block : blockGenerator) {
-      co_yield block;
-    }
-    auto lastBlock = getIncompleteBlock(endBlockMetadata - 1);
-    pruneBlock(lastBlock, limitOffset);
-    if (!lastBlock.empty()) {
-      details.numElementsYielded_ += lastBlock.numRows();
-      co_yield lastBlock;
+      if (beginBlockMetadata + 1 < endBlockMetadata) {
+        if (!middleBlocksYielded) {
+          // Get and yield the remaining blocks.
+          if (!generatorCreated) {
+            generatorCreated = true;
+            middleBlocksGenerator = reader->asyncParallelBlockGenerator(
+                beginBlockMetadata + 1, endBlockMetadata - 1, config,
+                cancellationHandle, limitOffset);
+            middleBlocksGenerator.setDetailsPointer(&details());
+          }
+
+          auto block{middleBlocksGenerator.get()};
+          if (block.has_value()) {
+            return std::move(block.value());
+          } else {
+            middleBlocksYielded = true;
+          }
+        }
+
+        if (!lastBlockYielded) {
+          lastBlockYielded = true;
+          auto block = getIncompleteBlock(endBlockMetadata - 1);
+          pruneBlock(block, limitOffset);
+          if (!block.empty()) {
+            details().numElementsYielded_ += block.numRows();
+            return block;
```

Comment:
add separate member functions

`yieldFirstBlock()` `setupGenerator` `yieldMiddleBlocks`, etc. , 
then you have the (simple) control flow inside the `get()` and the rest all is manageable (in particular the setting up of the generator currently is a little messy).

Reply:
yes, without the coroutines this code got complicated

Reply:
this [commit](https://github.com/ad-freiburg/qlever/pull/2236/commits/6a830e505e6e1045871775fdb760137debbe7594) made it much simpler